### ISSUE TITLE
[a11y] Drop zone has focusable off-screen element

### ISSUE THREAD

**Issue by dpersing** [Shopify/polaris-react#762]
## Issue summary

The Drop zone component replaces the default file upload method with a custom button or a drag-and-drop-only zone, which causes a few issues for keyboard-only and screen reader users:

- All versions of the component have a visually hidden `<input>` element, which is visually hidden but can receive focus and is read by screen readers. When the component has a custom button, focus goes to both the button and the hidden `<input>`. Screen reader users will hear all instances of the upload controls, and sighted keyboard-only users will not be able tell where focus is.
- When the native control is hidden but it is not replaced visually with another control, there does not appear to be a way to upload files with the keyboard.
- If an item has been added to the dropzone, the visible button is removed. Keyboard users may have difficulty adding multiple items.
- If items are added using the drag and drop or separate `<button>`, the visually hidden `<input>` reports that no files have been uploaded.

![Drop zone with no visible button](https://user-images.githubusercontent.com/1462085/42187559-565d6d12-7e05-11e8-996f-6cf804409210.png)

![Small-size Drop zone with no visible button](https://user-images.githubusercontent.com/1462085/42187572-62243248-7e05-11e8-951d-82dde71fc46c.png)

#### Visually hidden file control (HTML/CSS)

```
<span class="Polaris-VisuallyHidden"><input type="file" multiple="" autocomplete="off"></span>
```

```
.Polaris-VisuallyHidden {
    position: absolute!important;
    top: 0;
    clip: rect(1px,1px,1px,1px)!important;
    overflow: hidden!important;
    height: 1px!important;
    width: 1px!important;
    padding: 0!important;
    border: 0!important;
}
```

### Hidden button

1. Using any browser, navigate to the [Drop zone](https://polaris.shopify.com/components/actions/drop-zone#navigation) examples and <kbd>tab</kbd> through them.
2. Repeat the process using any supported browser/screen reader combination.
3. Using the button provided, add items to the dropzone.
4. Read the updated information in the dropzone.

### Expected

- Focus will not go to off-screen controls.
- Controls will be consistent.
- State will be clear.

### Actual

- Focus goes to an off-screen control.
- Visible controls are hidden after uploads are added.
- The visually hidden but readable off-screen control reports a different upload state than is reflected by the items.

## Recommendations

- Use `display: none` to hide the off-screen `<input>` so it cannot be accessed by users, or use the standard `<input>` and not the additional button.
- For the Drop zone options that provide no button by default, display a control on focus so keyboard-only users can easily understand how to use the component.
- Keep a visible, focusable control even after items are added so keyboard users can continue to add items.

## Specifications

- **Component:** Drop zone
- **WCAG SCs:** [2.4.7 Focus Visible (Level AA)](https://www.w3.org/TR/WCAG21/#focus-visible)
- **Testing done with:**
    - VoiceOver/Safari/macOS
    - VoiceOver/Safari/iOS
    - NVDA/Firfox/Windows 10
    - Keyboard

**Comment by dpersing** [comment]
Ran into this again while working on documentation, so moved it over from the old repo.

**Comment by dpersing** [comment]
Hi @AndrewMusgrave! Any updates on this issue?
